policy_parser.py

```diff
@@ -14,35 +14,107 @@ def __init__(self, msg=None, policy=None):
 
 
 class PolicyParser(object):
+    """
+    Parser for Kubernetes NetworkPolicy API objects.
+
+    They are defined by the NetworkPolicySpec in the /apis/networking API
+    group.
+    """
     def __init__(self, policy):
-        self.namespace = policy["metadata"]["namespace"]
+        """
+        Create a Parser for a Kubernetes NetworkPolicy API object.
+
+        Returns the parser.
+        Throws PolicyError if the Policy is invalid.
+        """
         self.policy = policy
+        self.namespace = self.policy["metadata"]["namespace"]
+
+    def calculate_pod_selector(self):
+        """
+        Generate the Calico pod selector for this Policy.
+
+        Returns the pod selector in the Calico datamodel format.
+        Throws PolicyError if the Policy is invalid or unsupported.
+        """
+        _log.debug("Calculating pod selector")
+        k8s_selector = self.policy["spec"]["podSelector"]
+
+        # PodSelectors only select pods from the Policy's namespace.
+        calico_selectors = ["%s == '%s'" % (K8S_NAMESPACE_LABEL, self.namespace)]
+
+        calico_selectors += self._calculate_selectors(k8s_selector)
+      
+        _log.debug("Selector with %d filters" % len(calico_selectors))
+        return " && ".join(calico_selectors)
 
     def calculate_inbound_rules(self):
         """
-        Takes a NetworkPolicy object from the API and returns a list of
-        Calico Rule objects which should be applied on ingress.
+        Generate Calico Rule objects for this Policy's ingress rules.
+
+        Returns a list of Calico datamodel Rules.
+        Throws PolicyError if the Policy is invalid or unsupported.
         """
         _log.debug("Calculating inbound rules")
         rules = []
 
-        # Iterate through and create the appropriate Calico Rule objects.
-        allow_incomings = self.policy["spec"].get("ingress") or []
-        _log.debug("Found %s ingress rules", len(allow_incomings))
-        for allow_incoming_clause in allow_incomings:
-            # Convert each allow_incoming_clause into one or more
-            # Calico Rule objects.
-            if allow_incoming_clause:
-                # Rule exists - parse it.
-                r = self._allow_incoming_to_rules(allow_incoming_clause)
-            else:
-                # An empty rule means allow all traffic.
-                r = [Rule(action="allow")]
-            rules.extend(r)
+        ingress_rules = self.policy["spec"].get("ingress")
+        if ingress_rules:
+            _log.debug("Got %d ingress rules: translating to Calico format",
+                       len(ingress_rules))
+            for ingress_rule in ingress_rules:
+                _log.debug("Processing ingress rule %s", ingress_rule)
+                if ingress_rule:
+                    # Convert ingress rule into Calico Rules.
+                    _log.debug("Adding rule %s", ingress_rule)
+                    rules.extend(self._allow_incoming_to_rules(ingress_rule))
```

Comment:
"allow incoming" is old terminology from a previous version of the API.

We should update the naming here... something like "parse_ingress_rule" maybe?